File: src/console/font.zig
const std = @import("std");
const constants = @import("./constants.zig");

const DefaultFontData = @embedFile("font.dat");

pub const FontGlyph = struct {
    data: [constants.SPRITE_PIXELS]u1,

    pub fn init() FontGlyph {
        return FontGlyph{ .data = std.mem.zeroes([constants.SPRITE_PIXELS]u1) };
    }
};

pub const FontData = struct {
    glyphs: [constants.FONT_GLYPH_COUNT]FontGlyph,

    pub fn init(gpa: std.mem.Allocator) !FontData {
        var ret = FontData{ .glyphs = .{FontGlyph.init()} ** constants.FONT_GLYPH_COUNT };

        //const fdat = std.mem.bytesAsSlice(u1, DefaultFontData);
        //const fdat: *const [2048 * 8]u1 = @ptrCast(DefaultFontData);
        var fdat = try std.array_list.Aligned(u1, null).initCapacity(gpa, 2048 * 8);
        defer fdat.deinit(gpa);

        // Copy data to fdat
        for (0..DefaultFontData.len) |sidx| {
            var val = DefaultFontData[sidx];
            for (0..8) |_| {
                const b = (val & 0b10000000) >> 7;
                try fdat.append(gpa, @intCast(b));
                val <<= 1;
            }
        }

        //std.debug.print("Length of DefaultFontData: {d}\n", .{DefaultFontData.len});
        //std.debug.print("Length of fdat: {d}\n", .{fdat.items.len});

        //std.debug.print("Dumping font:\n", .{});
        //
        //for (0..FONT_GLYPH_COUNT) |ch| {
        //    const index = ch * 64;
        //    std.debug.print("{x:>02}: ", .{ch});
        //    var count: i32 = 0;
        //    for (0..64) |bit| {
        //        if (count >= 8) {
        //            std.debug.print("\n    ", .{});
        //            count = 0;
        //        }
        //        std.debug.print("{d}", .{fdat.items[index + bit]});
        //        count += 1;
        //    }
        //    std.debug.print("\n", .{});
        //}
        //std.debug.print("\n", .{});

        for (0..constants.FONT_GLYPH_COUNT) |ch| {
            const index = ch * 64;
            for (index..index + 64, 0..64) |bitidx, gbitidx| {
                //std.debug.print("{d:>4}:{d:>6}:{d:>6}:{d:>6}, ", .{ ch, index, gbitidx, bitidx });
                ret.glyphs[ch].data[gbitidx] = fdat.items[bitidx]; // TODO: Investigate index out of bounds
            }
            //std.debug.print("\n", .{});
        }
        //std.debug.print("\n", .{});

        return ret;
    }
};

File: src/console/font.cpp
#include "console/font.hpp"
#include <array>
#include <bitset>
#include <format>
#include <iostream>

#ifndef PRINT
#    define PRINT(msg) std::cerr << (msg);
#endif

namespace dracon::console {

    static inline const std::array<u8, FONT_GLYPH_COUNT * SPRITE_PIX_HEIGHT> DEFAULT_FONT_DATA = {
#include "./font.dat"
    };

    auto BuildGlyph(const std::array<u8, FONT_GLYPH_COUNT * SPRITE_PIX_HEIGHT> & arr, std::size_t start_idx)
      -> std::bitset<SPRITE_PIXELS> {
        // PRINT(std::format("BuildGlyph {}: ", start_idx));
        std::bitset<SPRITE_PIXELS> ret{};
        for (std::size_t i = 0; i < 8; i++) {
            // PRINT(std::format("{}, ", i));
            auto b           = arr.at(start_idx + i);
            ret[(i * 8)]     = (b >> 7) & 1;
            ret[(i * 8) + 1] = (b >> 6) & 1;
            ret[(i * 8) + 2] = (b >> 5) & 1;
            ret[(i * 8) + 3] = (b >> 4) & 1;
            ret[(i * 8) + 4] = (b >> 3) & 1;
            ret[(i * 8) + 5] = (b >> 2) & 1;
            ret[(i * 8) + 6] = (b >> 1) & 1;
            ret[(i * 8) + 7] = (b >> 0) & 1;
        }
        // PRINT("\n");
        return ret;
    }

    auto FontData::LoadDefault() -> void {
        // PRINT("FontData::LoadDefault:\n");
        for (std::size_t i = 0; i < Glyphs.size(); i++) {
            // PRINT(std::format(" {}: ", i));
            auto gdat         = BuildGlyph(DEFAULT_FONT_DATA, i * (SPRITE_PIXELS / 8));
            Glyphs.at(i).Data = gdat;
        }
        // PRINT("\n");
    }

} // namespace dracon::console
